physics/SFC_Models/Land/Noahmp/noahmpdrv.F90

```diff
@@ -9,99 +9,446 @@
+    real(kind_phys), dimension(:), intent(in) :: xlon    ! longitude !GFS_Data(cdata%blk_no)%Grid%xlon
+    real(kind_phys), dimension(:), intent(in) :: xlat    ! latitude
+
+    integer,                       intent(in) :: lsoil, lsnow_lsm
+    real(kind=kind_phys),          intent(in) :: dtp, fhour
+
+    ! Initialize CCPP error handling variables
+    errmsg = ''
+    errflg = 0
+
+    ! Consistency checks
+    if (lsm/=lsm_noahmp) then
+      write(errmsg,'(*(a))') 'Logic error: namelist choice of ',   &
+    &       'LSM is different from Noah'
+      errflg = 1
+      return
+    end if
 
-        implicit none
-        integer,              intent(in) :: lsm
-        integer,              intent(in) :: lsm_noahmp    
-        integer,              intent(in)  :: me, isot, ivegsrc, nlunit
+    if (ivegsrc /= 1) then
+      errmsg = 'The NOAHMP LSM expects that the ivegsrc physics '// &
+                'namelist parameter is 1. Exiting...'
+      errflg = 1
+      return
+    end if
+    if (isot /= 1) then
+      errmsg = 'The NOAHMP LSM expects that the isot physics '// &
+                'namelist parameter is 1. Exiting...'
+      errflg = 1
+      return
+    end if
 
-        real (kind=kind_phys), dimension(:), intent(out) :: pores, resid
+    if ( do_mynnsfclay .and. .not. do_mynnedmf) then
+      errmsg = 'Problem : do_mynnsfclay = .true.' // &
+                'but mynnpbl is .false.. Exiting ...'
+      errflg = 1
+      return
+    end if
 
-        logical,              intent(in) :: do_mynnsfclay
-        logical,              intent(in) :: do_mynnedmf
+    !--- initialize soil vegetation
+    call set_soilveg(me, isot, ivegsrc, nlunit, errmsg, errflg)
 
+    !--- read in noahmp table
+    call read_mp_table_parameters(errmsg, errflg)
 
-        character(len=*),     intent(out) :: errmsg
-        integer,              intent(out) :: errflg
+    ! initialize psih and psim 
 
-        ! Initialize CCPP error handling variables
-        errmsg = ''
-        errflg = 0
+    if ( do_mynnsfclay ) then
+    call psi_init(psi_opt,errmsg,errflg)
+    endif
 
-        ! Consistency checks
-        if (lsm/=lsm_noahmp) then
-          write(errmsg,'(*(a))') 'Logic error: namelist choice of ',   &
-       &       'LSM is different from Noah'
-          errflg = 1
-          return
-        end if
+    pores (:) = maxsmc (:)
+    resid (:) = drysmc (:)
 
-        if (ivegsrc /= 1) then
-          errmsg = 'The NOAHMP LSM expects that the ivegsrc physics '// &
-                   'namelist parameter is 1. Exiting...'
-          errflg = 1
-          return
-        end if
-        if (isot /= 1) then
-          errmsg = 'The NOAHMP LSM expects that the isot physics '// &
-                   'namelist parameter is 1. Exiting...'
-          errflg = 1
-          return
-        end if
+    ! Read Land IAU settings 
+    call land_iau_mod_set_control(Land_IAU_Control, fn_nml, input_nml_file, &
```

Comment:
IMO, it's a little awkward that the control for this functionality isn't read in until the physics init time. This means that every physics suite that uses NoahMP will at least have to go into the `land_iau_mod_set_control` routine to read the namelist. It would probably be cleaner to at least read in the do_land_iau variable in the host so that the physics can avoid doing anything with land IAU when it's not being used. I'm not aware of other physics schemes that force the host to read a namelist within the physics in order to know whether an option is used or not. I know that there is a GWD scheme that has a separate nml, but I think that there is also a namelist variable in the "main" physics namelist that controls whether the scheme namelist is even read. @dustinswales Do you have an opinion here?

Reply:
@grantfirl I have now moved the "land_iau_mod_set_control" call to "sfcprop_create". The land IAU namelist is also read here. I considered reading the land IAU specs similar to that of the atm IAU (in GFS_Control), but the sfcprop_create occurs in close proximity to GFS_Control_init, so I thought it is better to keep all land IAU related namelist reading and setup in one place.